pylint/checkers/similar.py

```diff
@@ -26,32 +26,315 @@
 """
+import copy
 import functools
+import itertools
+import operator
+import random
 import re
 import sys
 from collections import defaultdict
 from getopt import getopt
 from io import TextIOWrapper
 from itertools import chain, groupby
-from typing import List
-
-import astroid
+from typing import (
+    Any,
+    Dict,
+    FrozenSet,
+    Generator,
+    Iterable,
+    List,
+    NamedTuple,
+    NewType,
+    Tuple,
+)
+
+import astroid  # type: ignore
 
 from pylint.checkers import BaseChecker, MapReduceMixin, table_lines_from_stats
 from pylint.interfaces import IRawChecker
 from pylint.reporters.ureports.nodes import Table
 from pylint.utils import decoding_stream
 
+DEFAULT_MIN_SIMILARITY_LINE = 4
+
 REGEX_FOR_LINES_WITH_CONTENT = re.compile(r".*\w+")
 
+# Index defines a location in a LineSet stripped lines collection
+Index = NewType("Index", int)
+
+# LineNumber defines a location in a LinesSet real lines collection (the whole file lines)
+LineNumber = NewType("LineNumber", int)
+
+
+# LineSpecifs holds characteristics of a line in a file
+class LineSpecifs(NamedTuple):
+    text: str
+    linenumber: LineNumber
+
+
+# Links LinesChunk object to the starting indices (in lineset's stripped lines)
+# of the different chunk of lines that are used to compute the hash
+HashToIndex_T = Dict["LinesChunk", List[Index]]
+
+# Links index in the lineset's stripped lines to the real lines in the file
+IndexToLines_T = Dict[Index, "SuccessiveLinesLimits"]
+
+
+class CplSuccessiveLinesLimits:
+    """
+    This class holds a couple of SuccessiveLinesLimits objects, one for each file compared,
+    and a counter on the number of common lines between both stripped lines collections extracted
+    from both files
+    """
+
+    def __init__(self, first_file, second_file, effective_cmn_lines_nb):
+        self.first_file: "SuccessiveLinesLimits" = first_file
+        self.second_file: "SuccessiveLinesLimits" = second_file
+        self.effective_cmn_lines_nb: int = effective_cmn_lines_nb
+
+
+# Links the indices ot the starting line in both lineset's stripped lines to
+# the start and end lines in both files
+CplIndexToCplLines_T = Dict["LineSetStartCouple", CplSuccessiveLinesLimits]
+
+
+class LinesChunk:
+    """
+    The LinesChunk object computes and stores the hash of some consecutive stripped lines of a lineset.
+    """
+
+    __slots__ = (
+        "_fileid",  # The name of the file from which the LinesChunk object is generated
+        "_index",  # The index in the stripped lines that is the starting of consecutive lines
+        "_hash",  # The hash of some consecutive lines
+    )
+
+    def __init__(self, fileid: str, num_line: int, *lines: Iterable[str]):
+        self._fileid: str = fileid
+        self._index: Index = Index(num_line)
+        self._hash: int = 0
+        for lin in lines:
+            if lin:
+                self._hash += hash(lin)
+            else:
+                # Avoid matching empty lines
```

Comment:
If I understand correctly each empty line has a specific hash ? Wouldn't that break the duplication check when we encounter an empty line ? Shouldn't the hash be based on the hash of previous line and the hash of next line instead of being random ?

Reply:
Thanks @Pierre-Sassoulas. In fact those lines are useless as every empty line is removed from the stripped lines collection.